hal/tpm_io_linux.c

```diff
@@ -226,73 +236,88 @@
+                spi.len = 1;
                 do {
                     /* Check for SPI ready */
-                    spi.len = 1;
                     size = ioctl(spiDev, SPI_IOC_MESSAGE(1), &spi);
-                    if (rxBuf[0] & TPM_TIS_READY_MASK)
-                        break;
-                } while (size == 1 && --timeout > 0);
+                } while (
+                    (size == 1) &&
+                    ((rxBuf[TPM_TIS_HEADER_SZ-1] & TPM_TIS_READY_MASK)==0) &&
+                    (--timeout > 0));
             #ifdef WOLFTPM_DEBUG_TIMEOUT
                 printf("SPI Ready Timeout %d\n", TPM_SPI_WAIT_RETRY - timeout);
             #endif
-                if (size == 1 && timeout > 0) {
-                    ret = TPM_RC_SUCCESS;
-                }
-            }
-            else {
-                ret = TPM_RC_SUCCESS;
+                if (size != 1 )
+                    ret = TPM_RC_FAILURE;
+                else if (timeout <= 0)
+                    ret = TPM_RC_FAILURE;  /* Timeout */
             }
 
+            /* Remainder of message */
             if (ret == TPM_RC_SUCCESS) {
-                /* Remainder of message */
+                spi.cs_change = 0;    /* Deassert cs after transfer */
                 spi.tx_buf   = (unsigned long)&txBuf[TPM_TIS_HEADER_SZ];
                 spi.rx_buf   = (unsigned long)&rxBuf[TPM_TIS_HEADER_SZ];
                 spi.len      = xferSz - TPM_TIS_HEADER_SZ;
                 size = ioctl(spiDev, SPI_IOC_MESSAGE(1), &spi);
+                if (size != (size_t)xferSz - TPM_TIS_HEADER_SZ)
+                    ret = TPM_RC_FAILURE;
+            }
 
-                if (size == (size_t)xferSz - TPM_TIS_HEADER_SZ)
-                    ret = TPM_RC_SUCCESS;
+            /* Send 1 byte dummy message to deassert cs if needed */
+            if (spi.cs_change == 1) {
+                spi.cs_change = 0;
+                spi.len = 1;
+                size = ioctl(spiDev, SPI_IOC_MESSAGE(1), &spi);
+                (void)size;  /* Ignore result */
             }
     #else
             /* Send Entire Message - no wait states */
             spi.tx_buf   = (unsigned long)txBuf;
             spi.rx_buf   = (unsigned long)rxBuf;
             spi.len      = xferSz;
             size = ioctl(spiDev, SPI_IOC_MESSAGE(1), &spi);
-            if (size == (size_t)xferSz)
-                ret = TPM_RC_SUCCESS;
+            if (size != (size_t)xferSz)
+                ret = TPM_RC_FAILURE;
     #endif /* WOLFTPM_CHECK_WAIT_STATE */
 
             close(spiDev);
         }
+        else {
+            /* Failed to open device */
+            ret = TPM_RC_FAILURE;
+        }
 
     #ifdef WOLFTPM_AUTODETECT
         /* if response is not 0xFF then we "found" something */
         if (!foundSpiDev) {
-            if (ret == TPM_RC_SUCCESS && rxBuf[0] != 0xFF) {
+            if ((ret == TPM_RC_SUCCESS) && (rxBuf[TPM_TIS_HEADER_SZ-1] != 0xFF)) {
```

Comment:
> 80 characters, also bonus parenthesis clutter things up, I think your only supposed to add them when you're doing bit wise operations

Reply:
Oops,  was gonna put that on 2 lines, so I added parenthesis to make it clearer.  Easier to do what you recommend.